From 2f0e32487d7ddf41237b6b76fcd3c05900aefff6 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Wed, 27 Apr 2022 15:17:39 +0800
Subject: [PATCH 77/98] backend-drm: Add master mode to control CRTC management

Stop disabling unused CRTCs by default to allow multiple Weston instances
to run on different monitors. Set WESTON_DRM_MASTER=1 to revert to
original behavior and manage all CRTCs.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 libweston/backend-drm/drm-internal.h | 3 +++
 libweston/backend-drm/drm.c          | 8 ++++++++
 libweston/backend-drm/kms.c          | 7 ++++---
 3 files changed, 15 insertions(+), 3 deletions(-)

diff --git a/libweston/backend-drm/drm-internal.h b/libweston/backend-drm/drm-internal.h
index bc8157c1..a1a30cd0 100644
--- a/libweston/backend-drm/drm-internal.h
+++ b/libweston/backend-drm/drm-internal.h
@@ -300,6 +300,9 @@ struct drm_backend {
 	/* Virtual screen size for resolution scaling */
 	int virtual_width;
 	int virtual_height;
+
+	/* Master mode flag, controls CRTC management */
+	bool master;
 };
 
 struct drm_mode {
diff --git a/libweston/backend-drm/drm.c b/libweston/backend-drm/drm.c
index a1668588..3932d91d 100644
--- a/libweston/backend-drm/drm.c
+++ b/libweston/backend-drm/drm.c
@@ -2915,6 +2915,10 @@ drm_output_detach_crtc(struct drm_output *output)
 
 	crtc->output = NULL;
 	output->crtc = NULL;
+
+	/* HACK: This is done here instead of in kms.c for the master mode */
+	drmModeSetCrtc(crtc->device->kms_device->fd,
+		       crtc->crtc_id, 0, 0, 0, NULL, 0, NULL);
 }
 
 static bool
@@ -4668,6 +4672,10 @@ drm_backend_create(struct weston_compositor *compositor,
 	if (buf)
 		sscanf(buf, "%dx%d", &b->virtual_width, &b->virtual_height);
 
+	buf = getenv("WESTON_DRM_MASTER");
+	if (buf && buf[0] == '1')
+		b->master = true;
+
 	wl_list_init(&b->kms_list);
 
 	b->compositor = compositor;
diff --git a/libweston/backend-drm/kms.c b/libweston/backend-drm/kms.c
index 56ab8c33..5317eaae 100644
--- a/libweston/backend-drm/kms.c
+++ b/libweston/backend-drm/kms.c
@@ -1635,7 +1635,7 @@ drm_pending_state_apply_atomic(struct drm_pending_state *pending_state,
 		break;
 	}
 
-	if (device->recovery_status == DRM_RECOVERY_SCHEDULED) {
+	if (b->master && device->recovery_status == DRM_RECOVERY_SCHEDULED) {
 		struct weston_head *head_base;
 		struct drm_head *head;
 		struct drm_crtc *crtc;
@@ -1834,7 +1834,7 @@ drm_pending_state_apply(struct drm_pending_state *pending_state)
 		return drm_pending_state_apply_atomic(pending_state,
 						      DRM_STATE_APPLY_ASYNC);
 
-	if (device->recovery_status == DRM_RECOVERY_SCHEDULED) {
+	if (b->master && device->recovery_status == DRM_RECOVERY_SCHEDULED) {
 		/* If we need to reset all our state (e.g. because we've
 		 * just started, or just been VT-switched in), explicitly
 		 * disable all the CRTCs we aren't using. This also disables
@@ -1903,6 +1903,7 @@ int
 drm_pending_state_apply_sync(struct drm_pending_state *pending_state)
 {
 	struct drm_device *device = pending_state->device;
+	struct drm_backend *b = device->backend;
 	struct drm_output_state *output_state, *tmp;
 	struct drm_crtc *crtc;
 	bool failed;
@@ -1911,7 +1912,7 @@ drm_pending_state_apply_sync(struct drm_pending_state *pending_state)
 		return drm_pending_state_apply_atomic(pending_state,
 						      DRM_STATE_APPLY_SYNC);
 
-	if (device->recovery_status == DRM_RECOVERY_SCHEDULED) {
+	if (b->master && device->recovery_status == DRM_RECOVERY_SCHEDULED) {
 		/* If we need to reset all our state (e.g. because we've
 		 * just started, or just been VT-switched in), explicitly
 		 * disable all the CRTCs we aren't using. This also disables
-- 
2.39.5

